### PR TITLE
3-level supports

### PR DESCRIPTION
**Pull request by ChenZhao44**
This PR contains the 3-level supports for Bloqade.

### CODE REVIEW

File: lib/BloqadeExpr/src/types.jl [hunk 1]
```diff
@@ -295,8 +350,8 @@ end
 const RabiTypes = Union{Nothing,SumOfX,SumOfXPhase}
```
Comment (Roger-luo) [inline lib/BloqadeExpr/src/types.jl:350]: need to add a `{D}` here

File: lib/BloqadeExpr/src/types.jl [hunk 2]
```diff
@@ -295,8 +350,8 @@ end
...
 const DetuningTypes = Union{Nothing,SumOfN}
```
Comment (Roger-luo) [inline lib/BloqadeExpr/src/types.jl:351]: same

**Review by Roger-luo — CHANGES_REQUESTED** [review]
test coverage seems to be dropped for a few related files, e.g `mat.jl`, need more tests before merging

**Comment by GiggleLiu** [comment]
Can we avoid using the name type parameter?

File: lib/BloqadeExpr/src/interface.jl [hunk 3]
```diff
@@ -108,6 +108,99 @@ function rydberg_h(atom_positions, C, Ω, ϕ, Δ)
...
+    end
+
+    !isnothing(detuning_term) && (detuning_term = -detuning_term)
+
+    rh3 = rydberg_term
+    for t in (rabi_term, detuning_term)
+        !isnothing(t) && (rh3 += t) 
+    end
+
+    return rh3
```
Comment (weinbe58) [inline lib/BloqadeExpr/src/interface.jl:177]: It looks like you're not using the `RydbergHamiltonian` struct. I don't think this is an issue now but when the Schema gets updated to include 3-level support we will need to change that.

**Review by weinbe58 — COMMENTED** [review]
Overall it looks good! My only comment is we should think about support for BloqadeSchema in the future. I can work on this with a separate PR once this is merge though. I have a good idea on how to do it.

File: lib/BloqadeExpr/src/interface.jl (hunk 3 above)
Comment (ChenZhao44) [reply lib/BloqadeExpr/src/interface.jl:177]: We could not use `RydbergHamiltonian` here without changing its defination. For example, the Rabi term in 2-level systems is `RabiType` while it is a sum of `RabiType` in 3-level systems. Maybe we should add a certain struct for the 3-level.

**Comment by ChenZhao44** [comment]
> Can we avoid using the name type parameter?

We could avoid that. But I want to extent current defination of types instead of defining new types. Is there any disadvantage using name type parameters?

**Comment by GiggleLiu** [comment]
> > Can we avoid using the name type parameter?
> 
> We could avoid that. But I want to extent current defination of types instead of defining new types. Is there any disadvantage using name type parameters?

I think the Type name itself and D already determines the operator. Can you explain why is this additional type paramter is nessesary?

e.g. SumOfZ{2, :hyperfine} is not a valid type right?

**Comment by ChenZhao44** [comment]
> > > Can we avoid using the name type parameter?
> > 
> > 
> > We could avoid that. But I want to extent current defination of types instead of defining new types. Is there any disadvantage using name type parameters?
> 
> I think the Type name itself and D already determines the operator. Can you explain why is this additional type paramter is nessesary?
> 
> e.g. SumOfZ{2, :hyperfine} is not a valid type right?

We need the name to distinguise operators in 3-level system. For example, XPhase{3, :hyperfine} applies on |0> and |1> while XPhase{3, :rydberg} applies on |1> and |r>. I don't want to define too much different types and copy-paste similar codes. Could you give me an example of potential disadvantages?

**Comment by GiggleLiu** [comment]
> > > > Can we avoid using the name type parameter?
> > > 
> > > 
> > > We could avoid that. But I want to extent current defination of types instead of defining new types. Is there any disadvantage using name type parameters?
> > 
> > 
> > I think the Type name itself and D already determines the operator. Can you explain why is this additional type paramter is nessesary?
> > e.g. SumOfZ{2, :hyperfine} is not a valid type right?
> 
> We need the name to distinguise operators in 3-level system. For example, XPhase{3, :hyperfine} applies on |0> and |1> while XPhase{3, :rydberg} applies on |1> and |r>. I don't want to define too much different types and copy-paste similar codes. Could you give me an example of potential disadvantages?

To me it is not intuitive. Since the second field does not apply in general case, then users need to learn the new convention.

Another abvious disadvantage is longer input, easier to  make a typo.

e.g. when using X01, X1r for hyperfine and Rydberg, then if you make a typo like X1z, the compiler errors you about type not defined. If you use type parameter name, X{3, :Rydberg}, which is a typo, but this new type is considered as valid.

What Rydberg pulse can do is a finite set of operations, so I think defining it case by case is fine. Or if you want to define a generic gate for any two levels, you should add a new field `levels`, which can be a length 2 tuple, e.g. (1, 2), rather than a type parameter. Because by using type parameters, you did not really reduce the number of types.

**Comment by Roger-luo** [comment]
TL;DR, I don't have a strong opinion on using separate types, but it should not be too short. And I disagree with the points you made here about typos, they have no big difference in the implementation. 

> Another abvious disadvantage is longer input, easier to make a typo.

This sounds bikeshedding, longer input is more explicit, I think having a typo is better than getting something wrong working.

> If you use type parameter name, X{3, :Rydberg}, which is a typo, but this new type is considered as valid.

This still errors? so it's not valid.

> What Rydberg pulse can do is a finite set of operations, so I think defining it case by case is fine. Or if you want to define a generic gate for any two levels, you should add a new field levels, which can be a length 2 tuple, e.g. (1, 2), rather than a type parameter. Because by using type parameters, you did not really reduce the number of types.

The real issue is Yao does not provide the abstraction of basis, where the Pauli operators are actually the same for both hyperfine and rydberg, except the basis is different. I don't think `levels` is a good idea since the current design has no compilation entries, which makes it harder for dispatching. So it has to be something in the type parameter to make it work easier with the current setup with Yao. We also make this a type parameter, but I don't see much difference here between `:hyperfine` and `(1, 2)` in the type parameter.

ultimately, the solution would be having a semantic block in the expression about levels, e.g

```julia
struct OnBasis{D, B <: Basis}
   op::AbstractBlock
   basis::B
end
```

where `Basis` can be something like the [QuantumOptics basis](https://github.com/qojulia/QuantumOpticsBase.jl/blob/master/src/spin.jl#L10), but inside a symbolic expression, this, however, will require us re-think the convention of blocks to not attach with a matrix (not all operators have matrices if we actually want a symbolic system for quantum operators) without a basis specified (we can always have a default basis on Pauli Z when calling `mat` for sure). Thus I don't think this change is gonna happen in short term.

**Comment by GiggleLiu** [comment]
> but I don't see much difference here between :hyperfine and (1, 2) in the type parameter.

Well, I mean as a field, rather than a type parameter.

**Comment by ChenZhao44** [comment]
What @GiggleLiu said make sense. Using name parameter could only get errors when inputing an invalid parameter during runtime while using seperated types could get errors during compile time. That is the biggest difference.

**Comment by Roger-luo** [comment]
> I do not think it is an issue about the basis. Rydberg atom is a qudit system already supported without Basis specification. Mixed systems might require a more general purposed basis specification.

there will be exponential many types of Pauli operators for the n-level system if there is no specification for basis in the expression. But you are right, this is not needed for 3-level at least for now.

> Mixed systems might require more general purposed basis specification.

yeah, that's something I'd like to consider in the rewrite of YaoBlocks, e.g operators on fock basis.

I think if @ChenZhao44 is willing to make the change then we are on the solution of using the two different types.

**Review by Roger-luo — APPROVED** [review]
I think `interface.jl` and `printing.jl` is effected by adding new types? maybe add some simple tests for those before merging
